launchers/launcher.go

```diff
@@ -199,14 +200,39 @@ func LaunchModules(ctx context.Context, modules config.ServerType) (context.Canc
 		}
 	}
 
-	// Include cache here just in case, although we currently don't use launcher to launch cache
+	// Origin needs to advertise once before the cache starts
+	if modules.IsEnabled(config.CacheType) && modules.IsEnabled(config.OriginType) {
+		log.Debug("Advertise Origin")
+		if err := server_ui.Advertise(ctx, servers); err != nil {
+			return shutdownCancel, err
+		}
+	}
+
+	if modules.IsEnabled(config.CacheType) {
+		// Give five seconds for the origin to finish advertising to the director
+		time.Sleep(5 * time.Second)
```

Comment:
Ah. Except this isn't waiting for the director to be up and running. The director already is up an running. This is waiting for the origin to finish its advertisement to the director. There's no easy way to check for that.

Reply:
What about querying the director at the endpoint that would redirect to that origin? Once the advertisement is complete, the director will respond with a 307 instead of a 404. I think `WaitUntilWorking` should be toolable for that.